File: sdk/ruby/test/wire/lists_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class ListsWireTest < WireMockTestCase
  def setup
    super

    @client = AgentMailer::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_lists_list_with_wiremock
    test_id = "lists.list.0"

    @client.lists.list(request_options: {
      additional_headers: {
        "X-Test-Id" => "lists.list.0"
      }
    })

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/lists",
      query_params: nil,
      expected: 1
    )
  end

  def test_lists_create_with_wiremock
    test_id = "lists.create.0"

    @client.lists.create(
      pod_id: "podId",
      name: "name",
      kind: "allow",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "lists.create.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/lists",
      query_params: nil,
      expected: 1
    )
  end

  def test_lists_delete_with_wiremock
    test_id = "lists.delete.0"

    @client.lists.delete(
      list_id: "listId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "lists.delete.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "DELETE",
      url_path: "/v1/lists/listId",
      query_params: nil,
      expected: 1
    )
  end

  def test_lists_add_entry_with_wiremock
    test_id = "lists.add_entry.0"

    @client.lists.add_entry(
      list_id: "listId",
      address_pattern: "addressPattern",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "lists.add_entry.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/lists/listId/entries",
      query_params: nil,
      expected: 1
    )
  end

  def test_lists_remove_entry_with_wiremock
    test_id = "lists.remove_entry.0"

    @client.lists.remove_entry(
      list_id: "listId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "lists.remove_entry.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "DELETE",
      url_path: "/v1/lists/listId/entries",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/lists_test.rs
use agentmailer::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_lists_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .lists
        .list(
            &ListsListQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/lists", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_lists_create_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .lists
        .create(
            &ListCreate {
                pod_id: "podId".to_string(),
                name: "name".to_string(),
                kind: ListCreateKind::Allow,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/lists", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_lists_delete_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client.lists.delete(&"listId".to_string(), None).await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("DELETE", "/v1/lists/listId", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_lists_add_entry_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .lists
        .add_entry(
            &"listId".to_string(),
            &ListEntry {
                address_pattern: "addressPattern".to_string(),
                note: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/lists/listId/entries", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_lists_remove_entry_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .lists
        .remove_entry(
            &"listId".to_string(),
            &RemoveEntryQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("DELETE", "/v1/lists/listId/entries", None, 1)
        .await
        .unwrap();
}
